### PR TITLE
Impl multilocation filter for asset-manager

### PR DESCRIPTION
**Pull request by Dengjianping**
Signed-off-by: Dengjianping <djptux@gmail.com>

## Description

closes: #XXXX

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Wrote unit tests.
- [ ] Updated relevant documentation in the code.
- [x] Added **one** line describing your change in `<branch>/CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] If runtime changes, need to update the version numbers properly:
   * `authoring_version`: The version of the authorship interface. An authoring node will not attempt to author blocks unless this is equal to its native runtime.
   * `spec_version`: The version of the runtime specification. A full node will not attempt to use its native runtime in substitute for the on-chain Wasm runtime unless all of spec_name, spec_version, and authoring_version are the same between Wasm and native.
   * `impl_version`: The version of the implementation of the specification. Nodes are free to ignore this; it serves only as an indication that the code is different; as long as the other two versions are the same then while the actual code may be different, it is nonetheless required to do the same thing. Non-consensus-breaking optimizations are about the only changes that could be made which would result in only the impl_version changing.
   * `transaction_version`: The version of the extrinsics interface. This number must be updated in the following circumstances: extrinsic parameters (number, order, or types) have been changed; extrinsics or pallets have been removed; or the pallet order in the construct_runtime! macro or extrinsic order in a pallet has been changed. If this number is updated, then the spec_version must also be updated 
- [ ] Verify benchmarks & weights have been updated for any modified runtime logics
- [ ] If import a new pallet, choose a proper module index for it, and allow it in `BaseFilter`. Ensure **every** extrinsic works from front-end. If there's corresponding tool, ensure both work for each other.
- [ ] If needed, update our Javascript/Typescript APIs. These APIs are offcially used by exchanges or community developers.
- [ ] If modifying existing runtime storage items, make sure to implement storage migrations for the runtime and test them with `try-runtime`. This includes migrations inhreited from upstream changes, and you can search the diffs for modifications of `#[pallet::storage]` items to check for any.

### CODE REVIEW

File: pallets/asset-manager/src/tests.rs [hunk 1]
```diff
@@ -189,3 +165,50 @@ fn update_asset() {
...
+
+#[test]
+fn filter_asset_location_should_work() {
+	let kusama_asset_metadata =
+		create_asset_metadata("Kusama", "KSM", 12, 1u128, None, false, false);
+	let kusama_location = AssetLocation(VersionedMultiLocation::V1(MultiLocation::parent()));
+
+	let manta_asset_metadata =
+		create_asset_metadata("Manta", "MANTA", 18, 1u128, None, false, false);
+	let mantan_location = AssetLocation(VersionedMultiLocation::V1(MultiLocation::new(
```
Comment (Garandor) [inline pallets/asset-manager/src/tests.rs:177]: just a typo -> mantan

File: pallets/asset-manager/src/tests.rs [hunk 2]
```diff
@@ -189,3 +165,50 @@ fn update_asset() {
...
+			Some(mantan_location.clone())
+		);
+
+		let new_location = AssetLocation(VersionedMultiLocation::V1(MultiLocation::new(
+			1,
+			X2(Parachain(1999), GeneralKey(b"UNKNOWN".to_vec())),
+		)));
+
+		// new location should be filtered
+		assert!(!crate::Pallet::<Runtime>::contains(&new_location));
```
Comment (Garandor) [inline pallets/asset-manager/src/tests.rs:210]: I'm not very familiar with the XCM code yet, but is this actually testing any "filtering"?

The way i'm reading this logic flow `new_location` is just a local variable that hasn't been given to `AssetManager::register_asset` yet (which presumably is where any filtering of foreign assets would happen), so the assert above will always be satisfied unless `new_location` happens to be a clone of `kusama_location` or `mantan_location`

I think I'm mainly confused by the "filtering" naming of the PR. Maybe you could add some description text to it to elaborate. The way I understand the code, the above test doesn't check filtering, it's only a sanity check/unit test for the `contains` implementation.

**Review by Garandor — CHANGES_REQUESTED** [review]
since you implemented contains on the pallet, maybe this
https://github.com/Manta-Network/Manta/blob/manta/pallets/asset-manager/src/lib.rs#L188 and
https://github.com/Manta-Network/Manta/blob/manta/pallets/asset-manager/src/lib.rs#L232
could be rewritten to Self.contains()

File: pallets/asset-manager/src/tests.rs (hunk 2 above)
Comment (stechu) [reply pallets/asset-manager/src/tests.rs:210]: filter means white-list. 
I agree the test should be in an end to end manner.

**Comment by stechu** [comment]
To test multilocation filter works, an end to end test to show that sending to location that not in the whitelist should return proper error.

**Comment by Dengjianping** [comment]
> 

I have to wait this PR #437 to be merged, because `xtokens` has this feature in 0.9.17.

**Comment by stechu** [comment]
> > 
> 
> I have to wait this PR #437 to be merged, because `xtokens` has this feature in 0.9.17.

Shall we move this out of 3.1.5 then?

**Comment by Dengjianping** [comment]
This feature will be implemented by #488 . I'm going to close it.

**Pull request closed without merging**